src/mudlet-lua/lua/geyser/GeyserButton.lua

```diff
@@ -0,0 +1,244 @@
+  me:setState(me.state)
+  me:resize() -- to pick up the Geyser.Button default size rather than Geyser.Label's
+  return me
+end
+
+--- Sets the state of the button to "up" or "down" and performs all accompanying visual transformations.
+-- Does not simulate the actual click, see Geyser.Button:press for that
+-- @param state the state to set the button to. One of "up" or "down"
+-- @see press
+function Geyser.Button:setState(state)
+  local stateType = type(state)
+  if stateType ~= "string" then
+    return nil, f"bad argument #1 type (state as string expected, got {stateType})"
+  end
+  state = state:lower()
+  if state ~= "up" and state ~= "down" then
+    return nil, f"bad argument #1 value (state must be one of 'up' or 'down', got {state})"
+  end
+  self.state = state
+  if state == "up" then
+    self:echo(self.msg)
+    if self.style then
+      if type(self.style) == "table" then
+        self:setStyleSheet(self.style:getCSS())
+      else
+        self:setStyleSheet(self.style)
+      end
+    else
+      self.parent.setColor(self, self.color)
+    end
+    self:setToolTip(self.tooltip, self.toolTipDuration)
+    return true
+  end
+  if not self.twoState then
+    return nil, "cannot set a single state button's state to 'down', only 'up'."
+  end
+  self:echo(self.downMsg)
+  if self.downStyle then
+    if type(self.downStyle) == "table" then
+      self:setStyleSheet(self.downStyle:getCSS())
+    else
+      self:setStyleSheet(self.downStyle)
+    end
+  else
+    self.parent.setColor(self, self.downColor)
+  end
+  self:setToolTip(self.downTooltip, self.toolTipDuration)
+end
+
+--- Handles clicking the button. If the button is twoState, also handles switching the button's state
+function Geyser.Button:press()
+  local command, func, newState
+  local state = self.state
+  if state == "up" then
+    command = self.clickCommand
+    func = self.clickFunction
+    newState = self.twoState and "down" or "up"
+  else
+    command = self.downCommand
+    func = self.downFunction
+    newState = "up"
+  end
+  if func then
+    local ok,err = pcall(func)
+    if not ok then
+      printError(err, true, true)
+    end
+  else
+    expandAlias(command)
+  end
+  self:setState(newState)
+end
+
+--- Stores the command to send when clicking the button in its 'down' state
+-- @param command the command to send. Will be run as though it is an alias.
+function Geyser.Button:setDownCommand(command)
+  local commandType = type(command)
+  if commandType ~= "string" then
+    printError(f"bad argument #1 type (command as string expected, got {commandType})", true, true)
+  end
+  self.downCommand = command
+end
+
+--- Stores the function to run when clicking the button in its 'down' state. Supercedes downCommand
+-- @param downFunction The function to run when the button is clicked in its 'down' state. Should be a lua function or valid lua codestring. 
+function Geyser.Button:setDownFunction(downFunction)
```

Comment:
Using this function seems to throw an error inside Geyser.Button,

`considerButton:setDownFunction("con")`

> [string "con"]:1: '=' expected near '' stack traceback: ..._Mudlet6tHisA/mudlet-lua/lua/geyser/GeyserButton.lua:146: in function 'setDownFunction'

Reply:
Aha... I would not have got that from the error message. Maybe more context in it would help?

Reply:
Yeah, I can add to the error message rather than just passing it along, that makes sense.

Reply:
ok, added something ahead which tells you what it was doing when it failed.